---
title: "TSPmap tutorial"
author: "J Grey Monroe"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---
Recent advances in genotyping by sequencing  (GBS) technologies have led to a dramatic increase in the number of markers used to generate genetic linkage maps. This can greatly improve the resolution for mapping quantitative traits to loci regulating ecologically and agriculturally important traits. However, these large marker datasets also increase the risk of misclustering and misordering of markers in maps generated with commonly used mapping software.  In order to address these issues, we created a new genetic mapping program, TSPmap, which uses a combination of approximate and exact Traveling Salesman Problem (TSP) solvers in an algorithm to handle large genomic marker datasets. We found that TSPmap can produce high quality linkage maps using a large number of markers and with relatively low sensitivity to missing and erroneous genotyping data compared to other methods. TSPmap provides a robust approach to efficiently generate high quality genetic maps with large numbers of homozygous genomic markers. 

# Installation

```{r, eval=FALSE}
install.packages("devtools")
library(devtools)
install_bitbucket("greymonroe/TSPmap")
```

Load package
```{r}
library(TSPmap)
```

# TSP solvers

TSPmap requires two existing tools to find the optimal solution:
LKH, http://www.akira.ruc.dk/~keld/research/LKH/
Concorde, http://www.math.uwaterloo.ca/tsp/concorde.html

Download the exceutable files approprirate for your system. The path to the each executable file must be specified in your code to call for functions within TSPmap:

```{r}
LKHexec = "~/TSPexecs/LKH"
Concordepath = "~/TSPexecs/concorde_mac"
```

# Create linkage map from marker data

Read in raw data. The code provided below loads a dataset that comes wtih the TSPmap package.
```{r}
rawdata = readFile(system.file("extdata", "JM20Demo_rawinput.tsv", package = "TSPmap"))
```

## Identify and remove duplicate markers.

```{r}
duplicates = finddups(rawdata, threshold = 99)
rawdata = removedups(rawdata, duplicates)
```

## Calculate recombination frequency between markers.
```{r}
rfmat = computeRFmat(rawdata, corrFlag=T)
```

## Create 'clean' recombination frequency matrix for more easy visualization (optional)
```{r}
rfmatclean = cleanrfmat(rfmat)
rfmatclean = addMarkerNames(rfmatclean, rawdata)
```

## Adjust recombination frequency values
Because it is detrimental to the performance of the TSP solver to allow these values to remain in the analysis, the user may input a cutoff threshold (default set to 0.4) above which all rf values are inflated to 0.5, preventing the solver from spending computation time optimizing these non-informative values.  However, setting the threshold too low can begin to affect the clustering and the final ordering of markers within the linkage group. This is especially true for noisy data sets, since meaningful rf values may be relatively high when there is a lot of noise in the data. 
```{r}
rfCap = 0.4
cappedrfmat = capRFmat(rfmat, rfCap, capval = 0.5)
```

## Set expected chromsome number
TSPmap also requires that users indicate the number of expected linkage groups, so it is helpful to know the chromosome count of the species being studied. 
```{r}
numChromosomes = 5
```

## Create raw clusters
Using the recombination frequency matrix, the unordered dataset is arranged in a minimum spanning tree and then broken into large clusters that may contain markers from more than one chromosome.

Next, each cluster is run through the approximate TSP solver, LKH, and checked for rf values that exceed a user-specified threshold. When such a value is detected, the cluster is broken apart to make sure that the final clusters do not contain markers from more than one chromosome.  This allows the user to control the sensitivity of the linkage group formation by setting what recombination frequency values may exist within a linkage group.
```{r}
rawclusters = autoClusterMST(rawdata, cappedrfmat, numChromosomes, LKHexec, internalRfThreshold=0.4)
```

## Merge clusters

Once the markers have been broken into well-separated clusters, the next step is to merge clusters that belong together in order to form the final linkage groups. 
```{r}
mergedclusters = autoMergeClusters(rawclusters, cappedrfmat, LKHexec, numChromosomes, rfCap, rawdata=rawdata)
```

## Order markers
The final clusters are processed using Concorde to produce the optimal linkage map for the given data.  
```{r}
finalMap = list()
for(i in 1:length(mergedclusters))
{
  finalMap[[i]] = createMap(rawdata, cappedrfmat, mergedclusters[i], Concordepath)
}
```
## Remove empty elements, if necessary
```{r}
finalMap = finalMap[!sapply(finalMap, is.null)]
```
## Convert final TSPmap solution to distance data.frame
```{r}
distancedf<-TSPmap2distancedf(finalMap)
```

## Reorder raw marker data
Reorders raw markerdata and adds linkage group and marker position based on TSPmap results and adds a numbered index column before genotype matrix. This produces the final product that can be used for downstreat analyses such as QTL mapping.
```{r}
markers<-reorderMarkers(rawdata = rawdata, distancedf = distancedf)
```
If saved as .csv can be imported to R/qtl by qtl:read.cross().
```{r, eval=F}
write.table(markers, file = "orderedMarkers.csv", sep=",", row.names = F)
Rqtlmap<-qtl::read.cross(file = "orderedMarkers.csv", format = "csv", genotypes = c("a","b" ,"-"))
```

## converts distancedf into table (optional)
Converts distancedf into table that can be converted into a map object by qtl::table2map
```{r, eval=T}
markertable<-distancedf2table(distancedf = distancedf)
Rqtlmap<-qtl::table2map(markertable)
plot(Rqtlmap)
```



